crates/ruff/src/rules/flake8_annotations/rules.rs

```diff
@@ -664,7 +664,7 @@ pub fn definition(
                             },
                             helpers::identifier_range(stmt, checker.locator),
                         );
-                        if checker.patch(diagnostic.kind.rule()) {
+                        if checker.patch((&diagnostic.kind).into()) {
```

Comment:
We could shorthand this by either (1) implementing more `From` variants (like `impl From<&Diagnostic> for &'static Rule`), or (2) letting `checker.patch` take `Diagnostic`.

Reply:
@konstin - Do you know of any pattern to make these conversions more streamlined? Or does this look right to you?

Reply:
i think `From` is the wrong trait for the kind of trick we want to do. i'd actually go with a `.rule()` on Diagnostic:

```rust
impl Diagnostic {
    pub fn rule(&self) -> &'static Rule {
        (&self.kind).into()
    }
}
```

otherwise this also works `diagnostic.kind.rule()`

```rust
        impl DiagnosticKind {
            pub fn rule(&self) -> &'static Rule {
                match self.rule.as_str() {
                    #from_impls_for_diagnostic_kind
                    _ => unreachable!("invalid rule name: {}", self.rule),
                }
            }
        }
```

Reply:
The issue is that we need to keep `Diagnostic` and `DiagnosticKind` separate from `Rule` -- they can't depend on `Rule`, otherwise all diagnostic implementations end up depending on all other implementations.

Specifically, `Rule` will end up being defined in the `ruff` crate, while `Diagnostic` and `DiagnosticKind` will probably be defined in some kind of `ruff_linter_base_types` crate (not the actual name, just an example).

Perhaps we could have an `AsRule` trait that we implement in the `ruff` crate for this conversion?

Reply:
Yeah, maybe a trait is the way to go here.

Reply:
Let me push a variant that uses that approach. It's actually far fewer changes.